### PR TITLE
TableSelector getAggregates optimization to only use PK columns when requesting COUNT(*) aggregate

### PR DESCRIPTION
**Pull request by cnathe**
#### Rationale
In the related PR, we made an optimization so that when generating the select SQL for a sample type, we skip the INNER JOIN to the materialized table when no columns from that table were being selected. This PR make the same type of change for Data Class tables. 

In addition, it makes a change to the TableSelector getAggregates() method so that if just the COUNT(*) aggregate is being requested, which happens for every DataRegion in LKS, we only use the PK columns for the given table. This allows for that aggregate query to make use of the optimizations made for the sample type and data class tables as mentioned above.

#### Related Pull Requests
* https://github.com/LabKey/platform/pull/4183

#### Changes
* Apply the checkSelectedColumnsFromRootOnly() check to Data Classes as well as Sample Types
* TableSelector getAggregates() method update to only use PK columns when getting just the COUNT(*) aggregate

### CODE REVIEW

File: api/src/org/labkey/api/data/TableSelector.java [hunk 1]
```diff
@@ -465,7 +465,12 @@ public boolean exists()
     // TODO: Convert to return Map<FieldKey, List<Aggregate.Result>>
     public Map<String, List<Result>> getAggregates(final List<Aggregate> aggregates)
     {
-        final AggregateSqlFactory sqlFactory = new AggregateSqlFactory(_filter, aggregates, _columns);
+        // If we are only asking for the COUNT(*) aggregate, then we only need the PK columns.
+        // This can make a big performance difference for Sample Type and Data Class tables as they can then skip
+        // the join between the exp schema base table and the materialized table for the given table.
+        Collection<ColumnInfo> aggColumns = aggregates.size() == 1 && aggregates.get(0).isCountStar() ? _table.getPkColumns() : _columns;
```
Comment (labkey-matthewb) [inline api/src/org/labkey/api/data/TableSelector.java]: Need to check that getPkColumns is non-empty().

**Review by labkey-matthewb — CHANGES_REQUESTED** [review]

File: api/src/org/labkey/api/data/TableSelector.java (hunk 1 above)
Comment (labkey-matthewb) [reply api/src/org/labkey/api/data/TableSelector.java]: It's also possible to do add an expr column like "new ExprColumn(table, "~one~", new SqlFragment("1"))

File: experiment/src/org/labkey/experiment/api/ExpRunItemTableImpl.java [hunk 2]
```diff
@@ -281,4 +281,29 @@ public void setDefaultVisibleColumns(@Nullable Iterable<FieldKey> list)
         super.setDefaultVisibleColumns(allowed);
     }
 
+    // if all selectedColumns are in the root table, no need to join to the provisioned table (e.g., in the COUNT(*) case)
+    protected boolean checkSelectedColumnsFromRootOnly(Set<FieldKey> selectedColumns, Set<String> rootColNames)
```
Comment (labkey-susanh) [inline experiment/src/org/labkey/experiment/api/ExpRunItemTableImpl.java]: I might change the name of this method to more clearly indicate what the true or false value it returns indicates. Something like `containsOnlyRootColumns`.

File: experiment/src/org/labkey/experiment/api/ExpDataClassDataTableImpl.java [hunk 3]
```diff
@@ -682,17 +684,20 @@ public SQLFragment getFromSQL(String alias)
...
+            }
         }
         sql.append(" FROM ");
         sql.append(_rootTable, "d");
-        sql.append(" INNER JOIN ").append(provisioned, "p").append(" ON d.lsid = p.lsid) ");
+        if (!skipProvisionedJoin)
+            sql.append(" INNER JOIN ").append(provisioned, "p").append(" ON d.lsid = p.lsid");
         String subAlias = alias + "_dc_sub";
-        sql.append(subAlias);
+        sql.append(") ").append(subAlias);
```
Comment (labkey-susanh) [inline experiment/src/org/labkey/experiment/api/ExpDataClassDataTableImpl.java:706]: Was this closing paren always missing? I'm not seeing where the open paren got added.

File: experiment/src/org/labkey/experiment/api/ExpRunItemTableImpl.java (hunk 2 above)
Comment (cnathe) [reply experiment/src/org/labkey/experiment/api/ExpRunItemTableImpl.java]: Good call. Changed.

File: experiment/src/org/labkey/experiment/api/ExpDataClassDataTableImpl.java (hunk 3 above)
Comment (cnathe) [reply experiment/src/org/labkey/experiment/api/ExpDataClassDataTableImpl.java:706]: Nope, the closing paren was always there. It was split out from end of the previous line 693.

File: api/src/org/labkey/api/data/TableSelector.java (hunk 1 above)
Comment (cnathe) [reply api/src/org/labkey/api/data/TableSelector.java]: Done. Thanks for the suggestion!

**Review by labkey-susanh — APPROVED** [review]

**Comment by labkey-matthewb** [comment]
Note for future possible changes:  We might also want to detect case when only provisioned columns are present.  It would be great to make the common lookup to sample type usages also avoid this join.  (with special case handling for LSID which can come from either table).

**Review by labkey-matthewb — APPROVED** [review]

**Pull request merged**